File: leetcode-java/src/main/java/com/melalex/leetcode/greedy/JumpGame7.java
package com.melalex.leetcode.greedy;

import java.util.BitSet;

public class JumpGame7 {

  public boolean canReach(String s, int minJump, int maxJump) {
    if (s.charAt(s.length() - 1) == '1') {
      return false;
    }

    BitSet canVisit = new BitSet();
    int max = 0;
    canVisit.set(0);

    for (int i = 0; i < s.length(); i++) {
      if (canVisit.get(i)) {
        int limit = Math.min(i + maxJump, s.length() - 1);

        for (int j = Math.max(i + minJump, max); j <= limit; j++) {
          canVisit.set(j, s.charAt(j) == '0');
        }

        max = limit;
      }
    }

    return canVisit.get(s.length() - 1);
  }
}

File: leetcode-scala/src/main/scala/com/leetcode/JumpGame7.scala
package com.leetcode

object JumpGame7 {

  import scala.collection.mutable

  def canReach(s: String, minJump: Int, maxJump: Int): Boolean = {
    if (s.last == '1')
      return false

    val couldLand = new mutable.BitSet()
    couldLand(0) = true

    for (i <- 0 until s.length) {
      if (couldLand(i)) {
        for (j <- (i + minJump) to Math.min(i + maxJump, s.length - 1)) {
          couldLand(j) = s(j) == '0'
        }
      }
    }

    couldLand(s.length - 1)
  }

  def main(args: Array[String]): Unit = {
    println(canReach("011000001011110000", 4, 7))
    println(canReach("011010", 2, 7))
  }
}
